### ISSUE TITLE
Move part of neighborlist calculation to the dataloader/dataset 

### ISSUE THREAD

**Issue by wiederm** [choderalab/modelforge#141]
The neighbor list calculations (calculating the local neighborhood N(i) for each i, the distance vector r_ij between atom pairs, and the distance d_ij) are repeated in each epoch. We should move this calculation to the dataloader and cache the results.

**Comment by wiederm** [comment]
A priori it isn't easy to know how much overhead the neighbor list calculation adds. To understand this better I passed a batch of data through the model and also performed the backward pass with the PyTorch Profiler. Below are the stack trace sorted (1) by total CPU time and (2) total GPU time. The neighborlist calculation are by far the most expensive calculations.

Sorted by total CPU time:


```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ---------------------------------------------------------------------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  Source Location                                                              
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ---------------------------------------------------------------------------  
                                            aten::index        23.65%      54.404ms        41.11%      94.565ms      30.040us      41.266ms        16.28%     101.261ms      32.167us          3148  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                           aten::arange         9.96%      22.908ms        13.90%      31.988ms      10.161us      18.577ms         7.33%      38.867ms      12.347us          3148  <built-in method arange of type object at 0x7f887bb60c20>                    
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                aten::repeat_interleave         8.30%      19.095ms        23.17%      53.311ms      67.740us      11.968ms         4.72%      54.457ms      69.196us           787  <built-in method repeat_interleave of type object at 0x7f887bb60c20>         
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                          aten::nonzero         8.24%      18.953ms         8.31%      19.125ms      12.151us      16.416ms         6.48%      22.545ms      14.323us          1574  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                              aten::cat         8.20%      18.862ms         8.22%      18.913ms       8.011us      24.118ms         9.51%      24.173ms      10.238us          2361  <built-in method cat of type object at 0x7f887bb60c20>                       
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                          aten::reshape         5.82%      13.383ms         5.84%      13.431ms       4.267us      13.770ms         5.43%      20.052ms       6.370us          3148  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                     aten::index_select         3.75%       8.627ms         3.75%       8.636ms       8.636ms       8.624ms         3.40%       8.639ms       8.639ms             1  <built-in method index_select of Tensor object at 0x7f87981deab0>            
                                                                                                                                                                                                     modelforge/potential/models.py(121): calculate_r_ij                          
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                            aten::clone         3.45%       7.933ms         6.03%      13.869ms      17.623us       5.879ms         2.32%      15.216ms      19.334us           787  <built-in method repeat_interleave of type object at 0x7f887bb60c20>         
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                  
```


(2) sorted by total CUDA time:

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ---------------------------------------------------------------------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  Source Location                                                              
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ---------------------------------------------------------------------------  
                                            aten::index        24.67%      53.833ms        43.21%      94.307ms      30.111us      41.152ms        17.04%     101.122ms      32.287us          3132  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                              aten::cat         9.20%      20.088ms         9.23%      20.140ms       8.574us      25.239ms        10.45%      25.297ms      10.769us          2349  <built-in method cat of type object at 0x7ff574b2dc20>                       
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                           aten::arange        10.37%      22.624ms        14.41%      31.446ms      10.040us      18.514ms         7.67%      38.257ms      12.215us          3132  <built-in method arange of type object at 0x7ff574b2dc20>                    
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                          aten::nonzero         8.40%      18.334ms         8.77%      19.133ms      12.218us      16.367ms         6.78%      22.622ms      14.446us          1566  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                          aten::reshape         6.16%      13.433ms         6.18%      13.482ms       4.305us      13.965ms         5.78%      20.019ms       6.392us          3132  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
                                                                                                                                                                                                                                                                                  
                                aten::repeat_interleave         8.79%      19.175ms        24.13%      52.654ms      67.246us      11.836ms         4.90%      53.776ms      68.679us           783  <built-in method repeat_interleave of type object at 0x7ff574b2dc20>         
                                                                                                                                                                                                     modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                                                                                                  
                                           aten::select         3.58%       7.818ms         3.59%       7.837ms       5.004us       9.179ms         3.80%      11.833ms       7.556us          1566  modelforge/potential/models.py(64): enumerate_all_pairs                      
                                                                                                                                                                                                     modelforge/potential/models.py(216): forward                                 
                                                                                                                                                                                                     torch/nn/modules/module.py(1534): _call_impl                                 
                                                                                                                                                                                                     nn.Module: Neighborlist_0                                                    
                                                                                                                                                                                                     modelforge/potential/models.py(516): prepare_inputs                          
```

**Comment by wiederm** [comment]
@chrisiacovella and @ArnNag , if you agree with that interpretation of these findings, I will open a PR to move the neighborlist to the dataloader.

**Comment by wiederm** [comment]
We have discussed this offline and we came to the conclusion that we need to investigate a bit more. I am documenting some of the experiments here, starting with optimizing the data loader:
Q: Does it matter if we set `pin_memory` to True (as described [here](https://lightning.ai/docs/pytorch/stable/advanced/speed.html#dataloaders)?
A: No, it has no impact on training performance.
Q: Does it matter if we double the number of workers?
A:  No impact on training performance.

**Comment by wiederm** [comment]
The PyTorch Lightning Profiler (simple) report provides a summary of the time spent in various parts of the training and validation process over 10 epochs, here is the raw data and I follow this up with a summary:

```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                                                                                                                                                         	|  Mean duration (s)	|  Num calls      	|  Total time (s) 	|  Percentage %   	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                                                                                                                                                          	|  -              	|  3068           	|  37.397         	|  100 %          	|
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|  run_training_epoch                                                                                                                                             	|  3.039          	|  10             	|  30.39          	|  81.265         	|
|  run_training_batch                                                                                                                                             	|  0.15677        	|  70             	|  10.974         	|  29.345         	|
|  [LightningModule]TrainingAdapter.optimizer_step                                                                                                                	|  0.15531        	|  70             	|  10.871         	|  29.07          	|
|  [Strategy]SingleDeviceStrategy.training_step                                                                                                                   	|  0.13967        	|  70             	|  9.7769         	|  26.144         	|
|  [_EvaluationLoop].val_next                                                                                                                                     	|  0.72823        	|  11             	|  8.0105         	|  21.42          	|
|  [_TrainingEpochLoop].train_dataloader_next                                                                                                                     	|  0.11178        	|  70             	|  7.8243         	|  20.922         	|
|  [Strategy]SingleDeviceStrategy.validation_step                                                                                                                 	|  0.18371        	|  11             	|  2.0208         	|  5.4036         	|
|  [Strategy]SingleDeviceStrategy.backward                                                                                                                        	|  0.013277       	|  70             	|  0.9294         	|  2.4852         	|
```
The total time taken for the entire profiling session is 37 seconds, which is 100% of the time. 
The majority of the time (81.265%) is spent running the training epochs; validation takes a significant portion of time, about 21.42% of the total time, with each call taking about 0.72823 seconds on average.

Training steps and optimizer steps are significant parts of the per-batch processing time. Data loading and validation steps also consume a considerable amount of time.

**Comment by wiederm** [comment]
I have been using the Pytorch Lightning Profilere to better understand the time cost of the indivicudal calls, here is a closer look at the `training_step`:

```
Profile stats for: [Strategy]SingleDeviceStrategy.training_step
         744833 function calls (736538 primitive calls) in 10.494 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       70    0.000    0.000   10.495    0.150 strategy.py:380(training_step)
       70    0.001    0.000   10.494    0.150 training.py:214(training_step)
       70    0.002    0.000   10.449    0.149 training.py:97(compute_loss)
       70    0.001    0.000   10.441    0.149 training.py:58(_get_energies)
       70    0.001    0.000   10.439    0.149 models.py:592(forward)
 3640/140    0.006    0.000   10.431    0.075 module.py:1528(_wrapped_call_impl)
 3640/140    0.016    0.000   10.431    0.075 module.py:1534(_call_impl)
       70    0.000    0.000    8.471    0.121 models.py:516(prepare_inputs)
       70    0.052    0.001    8.470    0.121 models.py:216(forward)
       70    4.630    0.066    8.361    0.119 models.py:64(enumerate_all_pairs)
       70    0.002    0.000    1.960    0.028 models.py:723(forward)
       70    0.014    0.000    1.925    0.028 schnet.py:180(_forward)
      210    0.540    0.003    1.631    0.008 schnet.py:270(forward)
   294990    1.389    0.000    1.389    0.000 {built-in method torch.cat}
    98330    1.102    0.000    1.102    0.000 {built-in method torch.repeat_interleave}
     1190    0.006    0.000    0.936    0.001 utils.py:338(forward)
```

A large funk of the cumulative time is spent in the `enumerate_all_pairs` function.

**Comment by chrisiacovella** [comment]
I did effectively the same profiling, but on the full qm9 dataset; given the small dataset I wonder if we are just capturing some of the overhead that would actually be very minor in a more realistic setting (running on my laptop with a mobile 4060)

For the training_step it is pretty clear that the enumerate_all_pairs is a significant calculation.  For val_next, my machine appears to be much faster that the timing above (even with the larger dataset, but I assume this is the dataloader being efficient). In val_next,  the dataloader takes the most of the time, this operation is pretty minimal in the context of a 4 minute epoch, considering it's less than a second.    

**training_step**

```
Profile stats for: [Strategy]SingleDeviceStrategy.training_step
         14132773 function calls (14039395 primitive calls) in 232.100 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      837    0.003    0.000  232.103    0.277 strategy.py:380(training_step)
      837    0.004    0.000  232.094    0.277 training.py:214(training_step)
      837    0.012    0.000  231.920    0.277 training.py:97(compute_loss)
      837    0.006    0.000  231.874    0.277 training.py:58(_get_energies)
      837    0.005    0.000  231.855    0.277 models.py:592(forward)
43524/1674    0.029    0.000  231.764    0.138 module.py:1528(_wrapped_call_impl)
43524/1674    0.062    0.000  231.762    0.138 module.py:1534(_call_impl)
      837    0.001    0.000  230.508    0.275 models.py:516(prepare_inputs)
      837    0.120    0.000  230.501    0.275 models.py:216(forward)
      837  146.985    0.176  230.310    0.275 models.py:64(enumerate_all_pairs)
  5780358   27.526    0.000   27.526    0.000 {built-in method torch.cat}
  1928462   21.975    0.000   21.975    0.000 {built-in method torch.tensor}
  3853572   16.989    0.000   16.989    0.000 {built-in method torch.arange}
  1926786   16.825    0.000   16.825    0.000 {built-in method torch.repeat_interleave}
      837    0.007    0.000    1.252    0.001 models.py:723(forward)
      837    0.030    0.000    0.980    0.001 schnet.py:180(_forward)
     2511    0.100    0.000    0.721    0.000 schnet.py:270(forward)
    14229    0.044    0.000    0.520    0.000 utils.py:338(forward)
     5859    0.011    0.000    0.486    0.000 container.py:215(forward)
    14229    0.297    0.000    0.297    0.000 {built-in method torch._C._nn.linear}

```

**val_next**

```
Profile stats for: [_EvaluationLoop].val_next
         181562 function calls (181140 primitive calls) in 0.558 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  527/107    0.000    0.000    0.558    0.005 {built-in method builtins.next}
      107    0.000    0.000    0.558    0.005 combined_loader.py:339(__next__)
      107    0.000    0.000    0.557    0.005 combined_loader.py:128(__next__)
      107    0.001    0.000    0.557    0.005 dataloader.py:626(__next__)
      107    0.001    0.000    0.551    0.005 dataloader.py:1299(_next_data)
      110    0.000    0.000    0.548    0.005 dataloader.py:1266(_get_data)
      110    0.000    0.000    0.548    0.005 dataloader.py:1120(_try_get_data)
      110    0.001    0.000    0.548    0.005 queues.py:98(get)
      110    0.000    0.000    0.306    0.003 connection.py:253(poll)
      110    0.000    0.000    0.305    0.003 connection.py:439(_poll)
      110    0.001    0.000    0.305    0.003 connection.py:933(wait)
      110    0.000    0.000    0.303    0.003 selectors.py:402(select)
      110    0.303    0.003    0.303    0.003 {method 'poll' of 'select.poll' objects}
      110    0.006    0.000    0.240    0.002 {built-in method _pickle.loads}
      880    0.003    0.000    0.225    0.000 reductions.py:494(rebuild_storage_fd)
      880    0.001    0.000    0.213    0.000 resource_sharer.py:55(detach)
      880    0.003    0.000    0.153    0.000 resource_sharer.py:81(get_connection)

```
looking at the data loader for the training loop, pretty minimal time overall, so I don't think this is where we need to focus our time.  

Clearly it is the recomputing of the neighborlist
```
Profile stats for: [_TrainingEpochLoop].train_dataloader_next
         1492641 function calls (1489300 primitive calls) in 1.689 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 4177/837    0.003    0.000    1.687    0.002 {built-in method builtins.next}
      837    0.002    0.000    1.686    0.002 combined_loader.py:339(__next__)
      837    0.003    0.000    1.680    0.002 combined_loader.py:72(__next__)
      837    0.005    0.000    1.676    0.002 dataloader.py:626(__next__)
      837    0.004    0.000    1.649    0.002 dataloader.py:1299(_next_data)
      837    0.001    0.000    1.617    0.002 dataloader.py:1266(_get_data)
      837    0.001    0.000    1.615    0.002 dataloader.py:1120(_try_get_data)

```

**Comment by chrisiacovella** [comment]
So digging a bit deeper, most of this slow down is coming from my attempt to improve the memory footprint of the pair enumerate in PR #134 .  While I tested this initially and it was only slightly slower, the speed was very much dependent on my test batch size and dataset size (I think it was a case where everything was too small and dominated by overhead).  Running the more realistic systems shows huge differences in efficiency. 

Obviously, in the original implementation, we were running out of memory for datasets with larger molecules and/or larger batch sizes (limiting what nodes we can run on), so we need a different solution.  Moving the pair computation to the data loader is certainly the best approach as it will save us from having to repeat this calculation. If this doesn't pan out the way we want for whatever reason, the bit code below seems to be a reasonable option.  It's a similar approach to the original code, but takes into account each molecule size (it will only enumerate all possible pairs in a molecule, not all possible pairs in a batch).  It is much more memory efficient than the original algorithm, but about 25% slower (clearly much faster than the for loop approach). 

Consider the time for a single epoch on the gpu using spice114 with schnet and a batch size of 256 (512 will cause me to run out of memory pn my laptop):

- original really big tensors: 388 s
- for loop batching (what I did in #134): ~2880 s
- generate smaller tensors based on size of each molecule: 578 s

This is the code for the last pair enumeration:

```
    def enumerate_all_pairs_new(self, atomic_subsystem_indices: torch.Tensor):
        """Compute all pairs of atoms and their distances.

        Parameters
        ----------
        atomic_subsystem_indices : torch.Tensor, shape (nr_atoms_per_systems)
            Atom indices to indicate which atoms belong to which molecule
        """
        # generate index grid
        n = len(atomic_subsystem_indices)

        # get device that passed tensors lives on, initialize on the same device
        device = atomic_subsystem_indices.device

        # atomic_subsystem_indices are numbered from 0 to n_molecules - 1
        # in all cases, so we can use bincount to get the number of atoms in each molecule
        # which we can use to construct smaller tensors
        repeats = torch.bincount(atomic_subsystem_indices)
        offsets = torch.cat(
            (torch.tensor([0], device=device), torch.cumsum(repeats, dim=0)[:-1])
        )

        i_indices = torch.cat(
            [
                torch.repeat_interleave(
                    torch.arange(o, o + r, device=device), repeats=r
                )
                for r, o in zip(repeats, offsets)
            ]
        )
        j_indices = torch.cat(
            [
                torch.cat([torch.arange(o, o + r, device=device) for _ in range(r)])
                for r, o in zip(repeats, offsets)
            ]
        )

        # filter pairs to only keep those belonging to the same molecule
        same_molecule_mask = (
            atomic_subsystem_indices[i_indices] == atomic_subsystem_indices[j_indices]
        )
        i_final_pairs_temp = i_indices[same_molecule_mask]
        j_final_pairs_temp = j_indices[same_molecule_mask]

        if self.only_unique_pairs:
            # filter out pairs that are not unique
            unique_pairs_mask = i_final_pairs_temp < j_final_pairs_temp
            i_final_pairs = i_final_pairs_temp[unique_pairs_mask]
            j_final_pairs = j_final_pairs_temp[unique_pairs_mask]
        else:
            # filter out identical values
            unique_pairs_mask = i_final_pairs_temp != j_final_pairs_temp
            i_final_pairs = i_final_pairs_temp[unique_pairs_mask]
            j_final_pairs = j_final_pairs_temp[unique_pairs_mask]

        # concatenate to form final (2, n_pairs) tensor
        pair_indices = torch.stack((i_final_pairs, j_final_pairs))

        return pair_indices.to(device)
```

**Comment by chrisiacovella** [comment]
I put the code above in PR #146 ; additionally added in an if statement to use the original algorithm if we are only doing a single molecule (no need to do additional looping for a single molecule).  This saves about 25% of the time.

**Comment by wiederm** [comment]
I think this has been sufficiently addressed in PR #146
